Base implementation of LibPerpetuals

sporejack opened the pull request:
# Motivation
Shared types and logic are best abstracted into their own module for various reasons. This is what `LibPerpetuals` does.

# Changes
 - Define `Order` type
 - Update the rest of the codebase to use this type instead of `Types.Order`
 - Remove `Types.Order`

sporejack committed:
Create `LibPerpetuals` library

This commit creates a Solidity file for the `LibPerpetuals` library.

sporejack committed:
Define `Side` enumerated type

This commit defines the type `Side`, which is an enum representing
sides of a Tracer market.

sporejack committed:
Define `Order` type

This commit defines the `Order` type, which represents an order in a
Tracer market.

sporejack committed:
Define `Signature` type

This commit defines the `Signature` type, which represents the output
of the ECDSA.

sporejack committed:
Define `LimitOrder` type

This commit defines the `LimitOrder` type, which represents a
user-submitted order.

sporejack committed:
Define `SignedLimitOrder` type

This commit defines the `SignedLimitOrder` type, which represents an
ECDSA-signed limit order.

sporejack committed:
Make `LibPerpetuals` an actual library

This commit actually defines a Solidity library for `LibPerpetuals`.

sporejack committed:
Modify `Order` definition

This commit changes the `Order` type in two ways:

 - Add the `market` field
 - Remove the `filled` field

sporejack committed:
Remove (now) redundant types from `LibPerpetuals`

This commit removes the following types from `LibPerpetuals`, as we've
now decided they belong elsewhere:

 - `Signature`
 - `LimitOrder`
 - `SignedLimitOrder`

sporejack committed:
Remove `Order` from `Types.sol`

This commit removes the `Order` type from `Types.sol` as it now lives
in `LibPerpetuals`.

sporejack committed:
Merge branch 'contracts-v2' into 245-base-for-libperp

CalabashSquash commented on contracts/Trader.sol:
```diff
@@ -20,7 +21,7 @@ contract Trader is ITrader {
     bytes32 private constant ORDER_TYPE =
         keccak256(
-            "Order(address maker,uint256 amount,int256 price,uint256 filled,bool side,uint256 expiration,uint256 creation,address targetTracer)"
+            "Order(address maker,uint256 amount,int256 price,uint256 filled,bool side,uint256 expiration,uint256 created,address market)"
         );
```
I guess this should also change `expiration` to `expires`

sporejack replied:
Consult 55421b0.

CalabashSquash reviewed (changes requested):
One small change

sporejack committed:
Fix additional compilation issues

This commit fixes additional compilation issues (*with these changes*)
by adding an additional function to the `ITrader` interface, amonst
other things.

sporejack committed:
Change `expiration` field to `expires`

This commit fixes a small issue with the Trader shim.

raymogg commented on contracts/Trader.sol:
```diff
@@ -70,31 +78,34 @@ contract Trader is ITrader {
             // increment nonce if filled
-            bool completeMaker = makeOrder.filled == makeOrder.amount;
-            bool completeTaker = takeOrder.filled == takeOrder.amount;
+            bool completeMaker = makeOrderFilled == makeOrder.amount;
```
These comparisons are comparing the previous filled amount (before this trade was executed).

They should compare `filled[id] == makeOrder.amount`

sporejack replied:
Consult 2c0820d.

raymogg commented on contracts/Trader.sol:
```diff
@@ -150,7 +160,7 @@ contract Trader is ITrader {
      * @return an EIP712 compliant hash (with headers) of the limit order
      */
-    function hashOrder(Types.Order memory order) public view override returns (bytes32) {
+    function hashOrder(Perpetuals.Order memory order) public view override returns (bytes32) {
```
Just raising that this function is pretty much the same as `Perpetual.orderId(order)`, however it uses `abi.encodePacked`. I know this is used for comparing EIP712 signatures so may not be worth using the `orderId` function, but leaving a comment to just raise the point.

I realise this is used by the EIP712 spec for signing so maybe it shouldn't and this should be kept as is here, either in case we change the ID implementation or if we don't want to use `abi.encodePacked` for the order id just worth raising.

sporejack replied:
Does EIP-712 mandate packed encoding?

raymogg replied:
@sporejack due to how the ethSignTypedData RPC call works, I am fairly certain it needs data to be encoded in packed form, so yeah it does.

Thats the main reason I would be pro keeping this function. Just wanted to raise it so we were all aware

sporejack replied:
@raymogg If it *is* mandated then yes, I'd agree in that we do need this function to remain.

sporejack committed:
Add `canMatch` function to `LibPerpetuals`

This commit adds the `canMatch` function to the `LibPerpetuals`
library. This function implements the logic which determines whether
two orders (along with their currently-filled amounts) are eligible to
be matched against each other.

sporejack committed:
Use `canMatch`

This commit replaces the previous match-checking logic with `canMatch`.

sporejack committed:
Use fresh filled state in `Trader.sol`

This commit uses fresh filled state of orders in `Trader.sol`

sporejack committed:
Merge branch 'contracts-v2' of github.com:tracer-protocol/tracer-protocol into 245-base-for-libperp
